README

More documentation can be found in the doc/ directory or at http://www.recoll.org


                               Recoll user manual

  Jean-Francois Dockes

   <jfd@recoll.org>

   Copyright (c) 2005-2015 Jean-Francois Dockes

   Permission is granted to copy, distribute and/or modify this document
   under the terms of the GNU Free Documentation License, Version 1.3 or any
   later version published by the Free Software Foundation; with no Invariant
   Sections, no Front-Cover Texts, and no Back-Cover Texts. A copy of the
   license can be found at the following location: GNU web site.

   This document introduces full text search notions and describes the
   installation and use of the Recoll application. This version describes
   Recoll 1.21.

     ----------------------------------------------------------------------

   Table of Contents

   1. Introduction

                1.1. Giving it a try

                1.2. Full text search

                1.3. Recoll overview

   2. Indexing

                2.1. Introduction

                             2.1.1. Indexing modes

                             2.1.2. Configurations, multiple indexes

                             2.1.3. Document types

                             2.1.4. Indexing failures

                             2.1.5. Recovery

                2.2. Index storage

                             2.2.1. Xapian index formats

                             2.2.2. Security aspects

                2.3. Index configuration

                             2.3.1. Multiple indexes

                             2.3.2. Index case and diacritics sensitivity

                             2.3.3. The index configuration GUI

                2.4. Indexing WEB pages you wisit

                2.5. Extended attributes data

                2.6. Importing external tags

                2.7. Periodic indexing

                             2.7.1. Running indexing

                             2.7.2. Using cron to automate indexing

                2.8. Real time indexing

                             2.8.1. Slowing down the reindexing rate for fast
                             changing files

   3. Searching

                3.1. Searching with the Qt graphical user interface

                             3.1.1. Simple search

                             3.1.2. The default result list

                             3.1.3. The result table

                             3.1.4. Running arbitrary commands on result
                             files (1.20 and later)

                             3.1.5. Displaying thumbnails

                             3.1.6. The preview window

                             3.1.7. The Query Fragments window

                             3.1.8. Complex/advanced search

                             3.1.9. The term explorer tool

                             3.1.10. Multiple indexes

                             3.1.11. Document history

                             3.1.12. Sorting search results and collapsing
                             duplicates

                             3.1.13. Search tips, shortcuts

                             3.1.14. Saving and restoring queries (1.21 and
                             later)

                             3.1.15. Customizing the search interface

                3.2. Searching with the KDE KIO slave

                             3.2.1. What's this

                             3.2.2. Searchable documents

                3.3. Searching on the command line

                3.4. Path translations

                3.5. The query language

                             3.5.1. Modifiers

                3.6. Search case and diacritics sensitivity

                3.7. Anchored searches and wildcards

                             3.7.1. More about wildcards

                             3.7.2. Anchored searches

                3.8. Desktop integration

                             3.8.1. Hotkeying recoll

                             3.8.2. The KDE Kicker Recoll applet

   4. Programming interface

                4.1. Writing a document input handler

                             4.1.1. Simple input handlers

                             4.1.2. "Multiple" handlers

                             4.1.3. Telling Recoll about the handler

                             4.1.4. Input handler HTML output

                             4.1.5. Page numbers

                4.2. Field data processing

                4.3. API

                             4.3.1. Interface elements

                             4.3.2. Python interface

   5. Installation and configuration

                5.1. Installing a binary copy

                5.2. Supporting packages

                5.3. Building from source

                             5.3.1. Prerequisites

                             5.3.2. Building

                             5.3.3. Installation

                5.4. Configuration overview

                             5.4.1. Environment variables

                             5.4.2. The main configuration file, recoll.conf

                             5.4.3. The fields file

                             5.4.4. The mimemap file

                             5.4.5. The mimeconf file

                             5.4.6. The mimeview file

                             5.4.7. The ptrans file

                             5.4.8. Examples of configuration adjustments

Chapter 1. Introduction

1.1. Giving it a try

   If you do not like reading manuals (who does?) but wish to give Recoll a
   try, just install the application and start the recoll graphical user
   interface (GUI), which will ask permission to index your home directory by
   default, allowing you to search immediately after indexing completes.

   Do not do this if your home directory contains a huge number of documents
   and you do not want to wait or are very short on disk space. In this case,
   you may first want to customize the configuration to restrict the indexed
   area (for the very impatient with a completed package install, from the
   recoll GUI: Preferences -> Indexing configuration, then adjust the Top
   directories section).

   Also be aware that you may need to install the appropriate supporting
   applications for document types that need them (for example antiword for
   Microsoft Word files).

1.2. Full text search

   Recoll is a full text search application. Full text search finds your data
   by content rather than by external attributes (like a file name). You
   specify words (terms) which should or should not appear in the text you
   are looking for, and receive in return a list of matching documents,
   ordered so that the most relevant documents will appear first.

   You do not need to remember in what file or email message you stored a
   given piece of information. You just ask for related terms, and the tool
   will return a list of documents where these terms are prominent, in a
   similar way to Internet search engines.

   Full text search applications try to determine which documents are most
   relevant to the search terms you provide. Computer algorithms for
   determining relevance can be very complex, and in general are inferior to
   the power of the human mind to rapidly determine relevance. The quality of
   relevance guessing is probably the most important aspect when evaluating a
   search application.

   In many cases, you are looking for all the forms of a word, including
   plurals, different tenses for a verb, or terms derived from the same root
   or stem (example: floor, floors, floored, flooring...). Queries are
   usually automatically expanded to all such related terms (words that
   reduce to the same stem). This can be prevented for searching for a
   specific form.

   Stemming, by itself, does not accommodate for misspellings or phonetic
   searches. A full text search application may also support this form of
   approximation. For example, a search for aliterattion returning no result
   may propose, depending on index contents, alliteration alteration
   alterations altercation as possible replacement terms.

1.3. Recoll overview

   Recoll uses the Xapian information retrieval library as its storage and
   retrieval engine. Xapian is a very mature package using a sophisticated
   probabilistic ranking model.

   The Xapian library manages an index database which describes where terms
   appear in your document files. It efficiently processes the complex
   queries which are produced by the Recoll query expansion mechanism, and is
   in charge of the all-important relevance computation task.

   Recoll provides the mechanisms and interface to get data into and out of
   the index. This includes translating the many possible document formats
   into pure text, handling term variations (using Xapian stemmers), and
   spelling approximations (using the aspell speller), interpreting user
   queries and presenting results.

   In a shorter way, Recoll does the dirty footwork, Xapian deals with the
   intelligent parts of the process.

   The Xapian index can be big (roughly the size of the original document
   set), but it is not a document archive. Recoll can only display documents
   that still exist at the place from which they were indexed. (Actually,
   there is a way to reconstruct a document from the information in the
   index, but the result is not nice, as all formatting, punctuation and
   capitalization are lost).

   Recoll stores all internal data in Unicode UTF-8 format, and it can index
   files of many types with different character sets, encodings, and
   languages into the same index. It can process documents embedded inside
   other documents (for example a pdf document stored inside a Zip archive
   sent as an email attachment...), down to an arbitrary depth.

   Stemming is the process by which Recoll reduces words to their radicals so
   that searching does not depend, for example, on a word being singular or
   plural (floor, floors), or on a verb tense (flooring, floored). Because
   the mechanisms used for stemming depend on the specific grammatical rules
   for each language, there is a separate Xapian stemmer module for most
   common languages where stemming makes sense.

   Recoll stores the unstemmed versions of terms in the main index and uses
   auxiliary databases for term expansion (one for each stemming language),
   which means that you can switch stemming languages between searches, or
   add a language without needing a full reindex.

   Storing documents written in different languages in the same index is
   possible, and commonly done. In this situation, you can specify several
   stemming languages for the index.

   Recoll currently makes no attempt at automatic language recognition, which
   means that the stemmer will sometimes be applied to terms from other
   languages with potentially strange results. In practise, even if this
   introduces possibilities of confusion, this approach has been proven quite
   useful, and it is much less cumbersome than separating your documents
   according to what language they are written in.

   Before version 1.18, Recoll stripped most accents and diacritics from
   terms, and converted them to lower case before either storing them in the
   index or searching for them. As a consequence, it was impossible to search
   for a particular capitalization of a term (US / us), or to discriminate
   two terms based on diacritics (sake / sake, mate / mate).

   As of version 1.18, Recoll can optionally store the raw terms, without
   accent stripping or case conversion. In this configuration, it is still
   possible (and most common) for a query to be insensitive to case and/or
   diacritics. Appropriate term expansions are performed before actually
   accessing the main index. This is described in more detail in the section
   about index case and diacritics sensitivity.

   Recoll has many parameters which define exactly what to index, and how to
   classify and decode the source documents. These are kept in configuration
   files. A default configuration is copied into a standard location (usually
   something like /usr/[local/]share/recoll/examples) during installation.
   The default values set by the configuration files in this directory may be
   overridden by values that you set inside your personal configuration,
   found by default in the .recoll sub-directory of your home directory. The
   default configuration will index your home directory with default
   parameters and should be sufficient for giving Recoll a try, but you may
   want to adjust it later, which can be done either by editing the text
   files or by using configuration menus in the recoll GUI. Some other
   parameters affecting only the recoll GUI are stored in the standard
   location defined by Qt.

   The indexing process is started automatically the first time you execute
   the recoll GUI. Indexing can also be performed by executing the
   recollindex command. Recoll indexing is multithreaded by default when
   appropriate hardware resources are available, and can perform in parallel
   multiple tasks among text extraction, segmentation and index updates.

   Searches are usually performed inside the recoll GUI, which has many
   options to help you find what you are looking for. However, there are
   other ways to perform Recoll searches: mostly a command line interface, a
   Python programming interface, a KDE KIO slave module, and Ubuntu Unity
   Lens (for older versions) or Scope (for current versions) modules.

Chapter 2. Indexing

2.1. Introduction

   Indexing is the process by which the set of documents is analyzed and the
   data entered into the database. Recoll indexing is normally incremental:
   documents will only be processed if they have been modified since the last
   run. On the first execution, all documents will need processing. A full
   index build can be forced later by specifying an option to the indexing
   command (recollindex -z or -Z).

   recollindex skips files which caused an error during a previous pass. This
   is a performance optimization, and a new behaviour in version 1.21 (failed
   files were always retried by previous versions). The command line option
   -k can be set to retry failed files, for example after updating a filter.

   The following sections give an overview of different aspects of the
   indexing processes and configuration, with links to detailed sections.

   Depending on your data, temporary files may be needed during indexing,
   some of them possibly quite big. You can use the RECOLL_TMPDIR or TMPDIR
   environment variables to determine where they are created (the default is
   to use /tmp). Using TMPDIR has the nice property that it may also be taken
   into account by auxiliary commands executed by recollindex.

  2.1.1. Indexing modes

   Recoll indexing can be performed along two different modes:

     o Periodic (or batch) indexing: indexing takes place at discrete times,
       by executing the recollindex command. The typical usage is to have a
       nightly indexing run programmed into your cron file.

     o Real time indexing: indexing takes place as soon as a file is created
       or changed. recollindex runs as a daemon and uses a file system
       alteration monitor such as inotify, Fam or Gamin to detect file
       changes.

   The choice between the two methods is mostly a matter of preference, and
   they can be combined by setting up multiple indexes (ie: use periodic
   indexing on a big documentation directory, and real time indexing on a
   small home directory). Monitoring a big file system tree can consume
   significant system resources.

   The choice of method and the parameters used can be configured from the
   recoll GUI: Preferences -> Indexing schedule

  2.1.2. Configurations, multiple indexes

   The parameters describing what is to be indexed and local preferences are
   defined in text files contained in a configuration directory.

   All parameters have defaults, defined in system-wide files.

   Without further configuration, Recoll will index all appropriate files
   from your home directory, with a reasonable set of defaults.

   A default personal configuration directory ($HOME/.recoll/) is created
   when a Recoll program is first executed. It is possible to create other
   configuration directories, and use them by setting the RECOLL_CONFDIR
   environment variable, or giving the -c option to any of the Recoll
   commands.

   In some cases, it may be interesting to index different areas of the file
   system to separate databases. You can do this by using multiple
   configuration directories, each indexing a file system area to a specific
   database. Typically, this would be done to separate personal and shared
   indexes, or to take advantage of the organization of your data to improve
   search precision.

   The generated indexes can be queried concurrently in a transparent manner.

   For index generation, multiple configurations are totally independent from
   each other. When multiple indexes need to be used for a single search,
   some parameters should be consistent among the configurations.

  2.1.3. Document types

   Recoll knows about quite a few different document types. The parameters
   for document types recognition and processing are set in configuration
   files.

   Most file types, like HTML or word processing files, only hold one
   document. Some file types, like email folders or zip archives, can hold
   many individually indexed documents, which may themselves be compound
   ones. Such hierarchies can go quite deep, and Recoll can process, for
   example, a LibreOffice document stored as an attachment to an email
   message inside an email folder archived in a zip file...

   Recoll indexing processes plain text, HTML, OpenDocument
   (Open/LibreOffice), email formats, and a few others internally.

   Other file types (ie: postscript, pdf, ms-word, rtf ...) need external
   applications for preprocessing. The list is in the installation section.
   After every indexing operation, Recoll updates a list of commands that
   would be needed for indexing existing files types. This list can be
   displayed by selecting the menu option File -> Show Missing Helpers in the
   recoll GUI. It is stored in the missing text file inside the configuration
   directory.

   By default, Recoll will try to index any file type that it has a way to
   read. This is sometimes not desirable, and there are ways to either
   exclude some types, or on the contrary to define a positive list of types
   to be indexed. In the latter case, any type not in the list will be
   ignored.

   Excluding types can be done by adding wildcard name patterns to the
   skippedNames list, which can be done from the GUI Index configuration
   menu. For versions 1.20 and later, you can alternatively set the
   excludedmimetypes list in the configuration file. This can be redefined
   for subdirectories.

   You can also define an exclusive list of MIME types to be indexed (no
   others will be indexed), by setting the indexedmimetypes configuration
   variable. Example:

 indexedmimetypes = text/html application/pdf
          

   It is possible to redefine this parameter for subdirectories. Example:

 [/path/to/my/dir]
 indexedmimetypes = application/pdf
          

   (When using sections like this, don't forget that they remain in effect
   until the end of the file or another section indicator).

   excludedmimetypes or indexedmimetypes, can be set either by editing the

ChangeLog
2010-02-02 15:33 +0100  Jean-Francois Dockes  <jfd@recoll.org>  (d11da0283f03 [tip])

	* src/common/textsplit.cpp, src/common/textsplit.h,
	src/query/plaintorich.cpp, src/query/recollq.cpp,
	src/query/wasatorcl.cpp, src/rcldb/rcldb.cpp,
	src/rcldb/searchdata.cpp, src/rcldb/stoplist.cpp,
	src/rcldb/stoplist.h:
	cosmetics: use derived class for actual splitter instead of callback

2010-02-02 10:24 +0100  Jean-Francois Dockes  <jfd@recoll.org>  (a8caf709bcd3)

	* src/qt4gui/rclmain.ui, src/qtgui/rclmain.ui,
	src/qtgui/rclmain_w.cpp:
	QT GUI: define accelerators for res list page movements

2010-02-02 08:20 +0100  Jean-Francois Dockes  <jfd@recoll.org>  (ec31e285a553)

	* src/qtgui/reslist.cpp, src/query/reslistpager.h:
	Qt GUI: ensure that new page size is taken into account ASAP (no
	need for restarting app)

2010-02-01 17:51 +0100  Jean-Francois Dockes  <jfd@recoll.org>  (db953bb94c7f)

	* src/qt4gui/rclmain.ui, src/qtgui/preview_w.cpp,
	src/qtgui/rclmain.ui, src/qtgui/rclmain_w.cpp,
	src/qtgui/rclmain_w.h:
	QT GUI: add fullscreen mode

2010-02-01 10:31 +0100  Jean-Francois Dockes  <jfd@recoll.org>  (1eda55ae3be9)

	* src/mk/manifest.txt: new file.
	* src/excludefile, src/makesrcdist.sh, src/mk/manifest.txt:
	Making a source dist: check the new list against old reference +
	other checks

2010-01-31 19:53 +0100  Israel G. Lugo  <Israel G. Lugo>  (74d4e25d43c2)

	* src/recollinstall.in:
	Install recollq and its manpage when in cmdline mode. Don't install
	the recoll.1 manpage when in cmdline mode.

2010-01-31 19:47 +0100  Jean-Francois Dockes  <jfd@recoll.org>  (c88b0ef40512)

	* src/common/autoconfig.h.in:
	use 3-arg version of ac_define as the 1-arg one is being obsoleted

2010-01-31 19:45 +0100  Jean-Francois Dockes  <jfd@recoll.org>  (1960435ccb68)

	* src/configure.ac:
	Dispense with the x11-monitoring when neither fam nor inotify are
	configured

2010-01-31 19:35 +0100  Jean-Francois Dockes  <jfd@recoll.org>  (08e6abfc5fdf)

	* src/configure.ac:
	use 3-arg version of ac_define as the 1-arg one is being obsoleted

2010-01-31 19:34 +0100  Jean-Francois Dockes  <jfd@recoll.org>  (c0add9dd8ad4)

	* website/download.html:
	none

2010-01-30 17:47 +0100  Jean-Francois Dockes  <jfd@recoll.org>  (5ed138ff2230)

	* src/qtgui/spell_w.cpp, src/qtgui/spell_w.h:
	QT GUI: fix small problems in newly native qt4 term expander

2010-01-30 17:31 +0100  Jean-Francois Dockes  <jfd@recoll.org>  (6ecf959a8e01)

	* src/qt4gui/spell.ui: new file.
	* src/qt4gui/recollmain.ui: deleted file.
	* .hgignore, src/qt4gui/recollmain.ui, src/qt4gui/spell.ui,
	src/qt4gui/uifrom3, src/qtgui/spell_w.cpp, src/qtgui/spell_w.h:
	QT GUI: converted the qt4 term expander dialog to native qt4

2010-01-30 14:09 +0100  Jean-Francois Dockes  <jfd@recoll.org>  (df8a91aaff88)

	* src/qt4gui/rclmain.ui: new file.
	* .hgignore, src/qt4gui/rclmain.ui, src/qt4gui/uifrom3,
	src/qtgui/confgui/confguiindex.h, src/qtgui/rclmain_w.cpp,
	src/qtgui/rclmain_w.h:
	Qt GUI: made the Qt4 main window native Qt4 (no more use of
	Q3MainWindow)

2010-01-30 08:23 +0100  Jean-Francois Dockes  <jfd@recoll.org>  (ed18703563b7)

	* .hgignore, src/ChangeLog, website/BUGS.html, website/devel.html,
	website/doc.html, website/download.html, website/index.html.en:
	none

2010-01-30 08:23 +0100  Jean-Francois Dockes  <jfd@recoll.org>  (22044a3b2e2c)

	* src/qtgui/rclmain_w.h:
	Qt 4.6.1 Uic bug: change qt version test from == to <= as bug still
	here in 4.6.2

2010-01-30 08:21 +0100  Jean-Francois Dockes  <jfd@recoll.org>  (b1cb8c664953)

	* src/common/autoconfig.h.in, src/configure, src/configure.ac,
	src/index/recollindex.cpp, src/utils/x11mon.cpp:
	Renamed WITHOUT_X11 to DISABLE_X11MON for clarification

2010-01-30 08:18 +0100  Israel G. Lugo  <Israel G. Lugo>  (be03b72e1258)

	* src/configure.ac:
	Rename option without-gui to disable-qtgui. New option disable-
	x11mon. Separate control of creation of the gui and X11 session
	monitoring.

2010-01-29 19:00 +0100  Jean-Francois Dockes  <jfd@recoll.org>  (d95c21312a15)

	* src/kde/kioslave/recoll/CMakeLists.txt:
	KIO slave: fixed CMakeList to configure Recoll with --enable-pic

2010-01-29 17:22 +0100  Jean-Francois Dockes  <jfd@recoll.org>  (266941720a99)

	* src/python/README.txt: new file.
	* src/configure, src/configure.ac, src/doc/user/usermanual.sgml,
	src/lib/Makefile, src/lib/mkMake, src/mk/commondefs,
	src/mk/localdefs.in, src/php/00README.txt, src/python/README.txt,
	src/python/recoll/setup.py:
	Implemented configure --enable-pic flag to build the main lib with
	position-independent objects. This avoids having to edit localdefs
	by hand to build the new php extension, and voids the need for the
	Python module to recompile Recoll source files.

2010-01-29 15:47 +0100  Jean-Francois Dockes  <jfd@recoll.org>  (69c42078b8d3)

	* src/php/00README.txt: new file.
	* src/php/00README.txt, src/php/recoll/make.sh,
	src/php/recoll/recoll.cpp:
	PHP extension by Wenqiang Song : make ready for external use.
	- added minimal doc
	- fixed build script to work around php/libtool issue
	- have the module default to Query Language (instead of AND)

2010-01-28 18:22 +0100  Jean-Francois Dockes  <jfd@recoll.org>  (45e7ec5e16c5)

	* .hgignore, website/usermanual/README-dir.txt: new file.
	* packaging/debian/changelog, packaging/debian/compat,
	packaging/debian/control, packaging/debian/copyright,
	packaging/debian/docs, packaging/debian/menu,
	packaging/debian/rules, packaging/debian/watch: deleted file.
	* .hgignore, packaging/debian/changelog, packaging/debian/compat,
	packaging/debian/control, packaging/debian/copyright,
	packaging/debian/docs, packaging/debian/menu,
	packaging/debian/rules, packaging/debian/watch, src/makesrcdist.sh,
	website/usermanual/README-dir.txt:
	svn->mercurial modifications

2010-01-28 16:13 +0000  convert-repo  <convert-repo>  (e85c82d42126)

	* .hgtags: new file.
	* .hgtags:
	update tags

2010-01-26 13:23 +0000  dockes  <dockes>  (c0cb63a2702a)


utils/md5.cpp
/*    $OpenBSD: md5.c,v 1.4 2014/12/28 10:04:35 tedu Exp $    */

/*
 * This code implements the MD5 message-digest algorithm.
 * The algorithm is due to Ron Rivest.    This code was
 * written by Colin Plumb in 1993, no copyright is claimed.
 * This code is in the public domain; do with it what you wish.
 *
 * Equivalent code is available from RSA Data Security, Inc.
 * This code has been tested against that, and is equivalent,
 * except that you don't need to include two pages of legalese
 * with every copy.
 *
 * To compute the message digest of a chunk of bytes, declare an
 * MD5Context structure, pass it to MD5Init, call MD5Update as
 * needed on buffers full of bytes, and then call MD5Final, which
 * will fill a supplied 16-byte array with the digest.
 */

#ifdef BUILDING_RECOLL
#include "autoconfig.h"
#else
#include "config.h"
#endif

#include "md5.h"

#include <cstring>

#define PUT_BIT_LE(i, cp, value) do {                   \
        (cp)[i] = (uint8_t)(((value) >> 8 * i) & 0xFF); \
    } while (0)

#define PUT_64BIT_LE(cp, value) do {            \
        PUT_BIT_LE(7, cp, value);               \
        PUT_BIT_LE(6, cp, value);               \
        PUT_BIT_LE(5, cp, value);               \
        PUT_BIT_LE(4, cp, value);               \
        PUT_BIT_LE(3, cp, value);               \
        PUT_BIT_LE(2, cp, value);               \
        PUT_BIT_LE(1, cp, value);               \
        PUT_BIT_LE(0, cp, value);               \
    } while (0)

#define PUT_32BIT_LE(cp, value) do {            \
        PUT_BIT_LE(3, cp, value);               \
        PUT_BIT_LE(2, cp, value);               \
        PUT_BIT_LE(1, cp, value);               \
        PUT_BIT_LE(0, cp, value);               \
    } while (0)

static constexpr uint8_t PADDING[MD5_BLOCK_LENGTH] = {
    0x80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
};

/*
 * Start MD5 accumulation.  Set bit count to 0 and buffer to mysterious
 * initialization constants.
 */
void
MD5Init(MD5_CTX *ctx)
{
    ctx->count = 0;
    ctx->state[0] = 0x67452301;
    ctx->state[1] = 0xefcdab89;
    ctx->state[2] = 0x98badcfe;
    ctx->state[3] = 0x10325476;
}

/*
 * Update context to reflect the concatenation of another buffer full
 * of bytes.
 */
void
MD5Update(MD5_CTX *ctx, const void *inputptr, size_t len)
{
    auto input = static_cast<const uint8_t *>(inputptr);
    size_t have, need;

    /* Check how many bytes we already have and how many more we need. */
    have = static_cast<size_t>((ctx->count >> 3) & (MD5_BLOCK_LENGTH - 1));
    need = MD5_BLOCK_LENGTH - have;

    /* Update bitcount */
    ctx->count += static_cast<uint64_t>(len) << 3;

    if (len >= need) {
        if (have != 0) {
            memcpy(ctx->buffer + have, input, need);
            MD5Transform(ctx->state, ctx->buffer);
            input += need;
            len -= need;
            have = 0;
        }

        /* Process data in MD5_BLOCK_LENGTH-byte chunks. */
        while (len >= MD5_BLOCK_LENGTH) {
            MD5Transform(ctx->state, input);
            input += MD5_BLOCK_LENGTH;
            len -= MD5_BLOCK_LENGTH;
        }
    }

    /* Handle any remaining bytes of data. */
    if (len != 0)
        memcpy(ctx->buffer + have, input, len);
}

/*
 * Final wrapup - pad to 64-byte boundary with the bit pattern
 * 1 0* (64-bit count of bits processed, MSB-first)
 */
void
MD5Final(unsigned char digest[MD5_DIGEST_LENGTH], MD5_CTX *ctx)
{
    uint8_t count[8];
    size_t padlen;
    int i;

    /* Convert count to 8 bytes in little endian order. */
    PUT_64BIT_LE(count, ctx->count);

    /* Pad out to 56 mod 64. */
    padlen = MD5_BLOCK_LENGTH -
        ((ctx->count >> 3) & (MD5_BLOCK_LENGTH - 1));
    if (padlen < 1 + 8)
        padlen += MD5_BLOCK_LENGTH;
    MD5Update(ctx, PADDING, padlen - 8);        /* padlen - 8 <= 64 */
    MD5Update(ctx, count, 8);

    for (i = 0; i < 4; i++)
        PUT_32BIT_LE(digest + i * 4, ctx->state[i]);
    memset(ctx, 0, sizeof(*ctx));    /* in case it's sensitive */
}


/* The four core functions - F1 is optimized somewhat */

/* #define F1(x, y, z) (x & y | ~x & z) */
#define F1(x, y, z) (z ^ (x & (y ^ z)))
#define F2(x, y, z) F1(z, x, y)
#define F3(x, y, z) (x ^ y ^ z)
#define F4(x, y, z) (y ^ (x | ~z))

/* This is the central step in the MD5 algorithm. */
#define MD5STEP(f, w, x, y, z, data, s)                         \
    ( w += f(x, y, z) + data,  w = w<<s | w>>(32-s),  w += x )

/*
 * The core of the MD5 algorithm, this alters an existing MD5 hash to
 * reflect the addition of 16 longwords of new data.  MD5Update blocks
 * the data and converts bytes into longwords for this routine.
 */
void
MD5Transform(uint32_t state[4], const uint8_t block[MD5_BLOCK_LENGTH])
{
    uint32_t a, b, c, d, in[MD5_BLOCK_LENGTH / 4];

#if BYTE_ORDER == LITTLE_ENDIAN
    memcpy(in, block, sizeof(in));
#else
    for (a = 0; a < MD5_BLOCK_LENGTH / 4; a++) {
        in[a] = (uint32_t)(
            (uint32_t)(block[a * 4 + 0]) |
            (uint32_t)(block[a * 4 + 1]) <<  8 |
            (uint32_t)(block[a * 4 + 2]) << 16 |
            (uint32_t)(block[a * 4 + 3]) << 24);
    }
#endif

    a = state[0];
    b = state[1];
    c = state[2];
    d = state[3];

    MD5STEP(F1, a, b, c, d, in[ 0] + 0xd76aa478,  7);
    MD5STEP(F1, d, a, b, c, in[ 1] + 0xe8c7b756, 12);
    MD5STEP(F1, c, d, a, b, in[ 2] + 0x242070db, 17);
    MD5STEP(F1, b, c, d, a, in[ 3] + 0xc1bdceee, 22);
    MD5STEP(F1, a, b, c, d, in[ 4] + 0xf57c0faf,  7);
    MD5STEP(F1, d, a, b, c, in[ 5] + 0x4787c62a, 12);
    MD5STEP(F1, c, d, a, b, in[ 6] + 0xa8304613, 17);
    MD5STEP(F1, b, c, d, a, in[ 7] + 0xfd469501, 22);
    MD5STEP(F1, a, b, c, d, in[ 8] + 0x698098d8,  7);
    MD5STEP(F1, d, a, b, c, in[ 9] + 0x8b44f7af, 12);
    MD5STEP(F1, c, d, a, b, in[10] + 0xffff5bb1, 17);
    MD5STEP(F1, b, c, d, a, in[11] + 0x895cd7be, 22);
    MD5STEP(F1, a, b, c, d, in[12] + 0x6b901122,  7);
    MD5STEP(F1, d, a, b, c, in[13] + 0xfd987193, 12);
    MD5STEP(F1, c, d, a, b, in[14] + 0xa679438e, 17);
    MD5STEP(F1, b, c, d, a, in[15] + 0x49b40821, 22);

    MD5STEP(F2, a, b, c, d, in[ 1] + 0xf61e2562,  5);
    MD5STEP(F2, d, a, b, c, in[ 6] + 0xc040b340,  9);
    MD5STEP(F2, c, d, a, b, in[11] + 0x265e5a51, 14);
    MD5STEP(F2, b, c, d, a, in[ 0] + 0xe9b6c7aa, 20);
    MD5STEP(F2, a, b, c, d, in[ 5] + 0xd62f105d,  5);
    MD5STEP(F2, d, a, b, c, in[10] + 0x02441453,  9);
    MD5STEP(F2, c, d, a, b, in[15] + 0xd8a1e681, 14);
    MD5STEP(F2, b, c, d, a, in[ 4] + 0xe7d3fbc8, 20);
    MD5STEP(F2, a, b, c, d, in[ 9] + 0x21e1cde6,  5);
    MD5STEP(F2, d, a, b, c, in[14] + 0xc33707d6,  9);
    MD5STEP(F2, c, d, a, b, in[ 3] + 0xf4d50d87, 14);
    MD5STEP(F2, b, c, d, a, in[ 8] + 0x455a14ed, 20);
    MD5STEP(F2, a, b, c, d, in[13] + 0xa9e3e905,  5);
    MD5STEP(F2, d, a, b, c, in[ 2] + 0xfcefa3f8,  9);
    MD5STEP(F2, c, d, a, b, in[ 7] + 0x676f02d9, 14);
    MD5STEP(F2, b, c, d, a, in[12] + 0x8d2a4c8a, 20);

    MD5STEP(F3, a, b, c, d, in[ 5] + 0xfffa3942,  4);
    MD5STEP(F3, d, a, b, c, in[ 8] + 0x8771f681, 11);
    MD5STEP(F3, c, d, a, b, in[11] + 0x6d9d6122, 16);
    MD5STEP(F3, b, c, d, a, in[14] + 0xfde5380c, 23);
    MD5STEP(F3, a, b, c, d, in[ 1] + 0xa4beea44,  4);
    MD5STEP(F3, d, a, b, c, in[ 4] + 0x4bdecfa9, 11);
    MD5STEP(F3, c, d, a, b, in[ 7] + 0xf6bb4b60, 16);
    MD5STEP(F3, b, c, d, a, in[10] + 0xbebfbc70, 23);
    MD5STEP(F3, a, b, c, d, in[13] + 0x289b7ec6,  4);
    MD5STEP(F3, d, a, b, c, in[ 0] + 0xeaa127fa, 11);
    MD5STEP(F3, c, d, a, b, in[ 3] + 0xd4ef3085, 16);
    MD5STEP(F3, b, c, d, a, in[ 6] + 0x04881d05, 23);
    MD5STEP(F3, a, b, c, d, in[ 9] + 0xd9d4d039,  4);
    MD5STEP(F3, d, a, b, c, in[12] + 0xe6db99e5, 11);
    MD5STEP(F3, c, d, a, b, in[15] + 0x1fa27cf8, 16);
    MD5STEP(F3, b, c, d, a, in[2 ] + 0xc4ac5665, 23);

    MD5STEP(F4, a, b, c, d, in[ 0] + 0xf4292244,  6);
    MD5STEP(F4, d, a, b, c, in[7 ] + 0x432aff97, 10);
    MD5STEP(F4, c, d, a, b, in[14] + 0xab9423a7, 15);
    MD5STEP(F4, b, c, d, a, in[5 ] + 0xfc93a039, 21);
    MD5STEP(F4, a, b, c, d, in[12] + 0x655b59c3,  6);
    MD5STEP(F4, d, a, b, c, in[3 ] + 0x8f0ccc92, 10);
    MD5STEP(F4, c, d, a, b, in[10] + 0xffeff47d, 15);
    MD5STEP(F4, b, c, d, a, in[1 ] + 0x85845dd1, 21);
    MD5STEP(F4, a, b, c, d, in[8 ] + 0x6fa87e4f,  6);
    MD5STEP(F4, d, a, b, c, in[15] + 0xfe2ce6e0, 10);
    MD5STEP(F4, c, d, a, b, in[6 ] + 0xa3014314, 15);
    MD5STEP(F4, b, c, d, a, in[13] + 0x4e0811a1, 21);
    MD5STEP(F4, a, b, c, d, in[4 ] + 0xf7537e82,  6);
    MD5STEP(F4, d, a, b, c, in[11] + 0xbd3af235, 10);
    MD5STEP(F4, c, d, a, b, in[2 ] + 0x2ad7d2bb, 15);
    MD5STEP(F4, b, c, d, a, in[9 ] + 0xeb86d391, 21);

    state[0] += a;
    state[1] += b;
    state[2] += c;
    state[3] += d;
}

// C++ utilities
using std::string;

void MD5Final(string &digest, MD5_CTX *context)
{
    unsigned char d[16];
    MD5Final (d, context);
    digest.assign(reinterpret_cast<const char*>(d), 16);
}

string& MD5String(const string& data, string& digest)
{
    MD5_CTX ctx;
    MD5Init(&ctx);
    MD5Update(&ctx, reinterpret_cast<const unsigned char*>(data.c_str()), data.length());
    MD5Final(digest, &ctx);
    return digest;
}

string& MD5HexPrint(const string& digest, string &out)
{
    out.erase();
    out.reserve(33);
    static const char hex[]="0123456789abcdef";
    auto hash = reinterpret_cast<const unsigned char*>(digest.c_str());
    for (int i = 0; i < 16; i++) {
        out.append(1, hex[hash[i] >> 4]);
        out.append(1, hex[hash[i] & 0x0f]);
    }
    return out;
}

std::string MD5Hex(const std::string& data)
{
    std::string digest, out;
    MD5String(data, digest);
    MD5HexPrint(digest, out);
    return out;
}

string& MD5HexScan(const string& xdigest, string& digest)
{
    digest.erase();
    if (xdigest.length() != 32) {
        return digest;
    }
    for (unsigned int i = 0; i < 16; i++) {
        unsigned int val;
        if (sscanf(xdigest.c_str() + 2*i, "%2x", &val) != 1) {
            digest.erase();
            return digest;
        }
        digest.append(1, static_cast<unsigned char>(val));
    }
    return digest;
}
